DeepSpeech.ipynb

```diff
@@ -1021,80 +1170,132 @@
-    "        log_grads_and_vars(avg_tower_gradients)\n",
-    "        \n",
-    "        # Apply gradients to modify the model\n",
-    "        apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)\n",
-    "        \n",
-    "        # Create a saver to checkpoint the model\n",
-    "        saver = tf.train.Saver(tf.all_variables())\n",
+    "def train(session, data_sets):\n",
+    "    # Set n_steps parameter\n",
+    "    n_steps = data_sets.train.max_batch_seq_len\n",
+    "\n",
+    "    # Calculate the total number of batches\n",
+    "    total_batch = int(data_sets.train.num_examples/batch_size)\n",
+    "\n",
+    "    # Require that we have at least as many batches as devices\n",
+    "    assert total_batch >= len(available_devices)\n",
+    "\n",
+    "    # Create optimizer\n",
+    "    optimizer = create_optimizer()\n",
+    "\n",
+    "    # Get gradients for each tower (Runs across all GPU's)\n",
+    "    tower_decodings, tower_labels, tower_gradients, tower_loss, accuracy = \\\n",
+    "        get_tower_results(n_steps, data_sets.train, optimizer)\n",
+    "    \n",
+    "    # Validation step preparation\n",
+    "    validation_tower_decodings, validation_tower_labels = forward(session, data_sets.validation)\n",
+    "\n",
+    "    # Average tower gradients\n",
+    "    avg_tower_gradients = average_gradients(tower_gradients)\n",
+    "\n",
+    "    # Add logging of averaged gradients\n",
+    "    log_grads_and_vars(avg_tower_gradients)\n",
+    "\n",
+    "    # Apply gradients to modify the model\n",
+    "    apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)\n",
+    "\n",
+    "    # Create a saver to checkpoint the model\n",
+    "    saver = tf.train.Saver(tf.all_variables())\n",
+    "\n",
+    "    # Prepare tensor board logging\n",
+    "    merged = tf.merge_all_summaries()\n",
+    "    writer = tf.train.SummaryWriter(log_dir, session.graph)\n",
+    "\n",
+    "    # Init all variables in session\n",
+    "    session.run(tf.initialize_all_variables())\n",
+    "    \n",
+    "    # Init recent word error rate levels\n",
+    "    last_train_wer = 0.0\n",
+    "    last_validation_wer = 0.0\n",
+    "    \n",
+    "    # Loop over the data set for training_epochs epochs\n",
+    "    for epoch in range(training_iters):\n",
+    "        # Define total accuracy for the epoch\n",
+    "        total_accuracy = 0\n",
     "        \n",
-    "        # Create session in which to execute\n",
-    "        session = tf.Session(config=tf.ConfigProto(allow_soft_placement=True, log_device_placement=True))\n",
+    "        # Validation step to determine the best point in time to stop\n",
+    "        if epoch % validation_step == 0:\n",
+    "            _, last_validation_wer = print_wer_report(session, \"Validation\", validation_tower_decodings, validation_tower_labels)\n",
+    "            # TODO: Determine on base of WER, if model starts overfitting\n",
```

Comment:
> Determine on base of WER, if model starts overfitting

I don't think this makes sense. 

I read this as saying: If the model starts overfitting, then stop training only once the validation WER is below the target threshold.

Overfitting can occur and the validation WER could be above the threshold set and it can stay there. So, this could lead to an infinite loop.

We don't have to do this now, but we could use early stopping using the ValidationMonitor[[1]](https://www.tensorflow.org/versions/r0.10/tutorials/monitors/index.html#early-stopping-with-validationmonitor) and base early stopping on WER.